### PR TITLE
:sparkles: Migration wave tables latest and Jira Instance

### PR DESCRIPTION
**Pull request by gildub**
Migration Waves:
- Applications, Stakeholders and Tickets adjustable compound details 
- Activated trash can button to remove application from migration wave
- Follow hub changes now using `Ticket`'s parent.id and kind.id instead of names
- Rationalize/consolidate toastr.success/fail messages
- Move all components and forms under ./components
- Add "No Issues" for aggregated status to cover not documented case when migration wave has applications assigned but no tickets defined yet (not exported).

Jira Configuration:
- Enforce issue manager instance can't be delete when tickets exist with alert modal

Application form: 
- When updating application, added migrationwave field to preserve record otherwise application gets detached from migration wave.

Applications Inventory:
- Detail drawer includes either migration wave information

Queries/Model:
- The Id field of a resource exist for every resource record but at creation time.
  Removing it for being an optional field (`id?: number`) makes handling data much easier because there is no need to test for undefined across all resources. The only case where it's not needed is during creation time and this is where `Omit` is handy.
 Because the migration waves page is consolidating so many resources with Ref[] relationship it was obvious the Ref interface had to be changed.
Which impacted stakeholder, stakeholderGroup, tag, tagCategory changes and triggered a cascade of changes along the way including aligning the involved react query and rest queries.
The final solution (thanks to @mturley) is to use a generic type for all create REST function objects which opts out the id.
`export New<T extends {id: string"} = Omit<T, "id">` 

Resolves https://github.com/konveyor/tackle2-ui/issues/875, https://github.com/konveyor/tackle2-ui/issues/587, https://github.com/konveyor/tackle2-ui/issues/874, https://github.com/konveyor/tackle2-ui/issues/878, https://github.com/konveyor/tackle2-ui/issues/877

image

image

image

image

### CODE REVIEW

File: client/src/app/api/rest.ts [hunk 1]
```diff
@@ -227,50 +161,20 @@ export const getJobFunctions = (): AxiosPromise<JobFunction[]> => {
   return APIClient.get(`${JOB_FUNCTIONS}`, jsonHeaders);
 };
 
-// Tag categories
-
-export const getTagCategories = (): AxiosPromise<Array<TagCategory>> => {
```
Comment (ibolton336) [inline client/src/app/api/rest.ts:232]: Can you move this work to a separate PR?

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx [hunk 2]
```diff
@@ -109,6 +109,55 @@ export const StakeholderGroupForm: React.FC<StakeholderGroupFormProps> = ({
...
+    _: AxiosResponse<StakeholderGroup>
+  ) =>
+    pushNotification({
+      title: t("toastr.success.create", {
+        type: t("terms.stakeholderGroup"),
+      }),
+      variant: "success",
+    });
+
+  const onCreateStakeholderGroupError = (error: AxiosError) => {
```
Comment (ibolton336) [inline client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:122]: Can you move all unrelated work to new PRs with a description & intention behind the changes? Thanks.

**Review by ibolton336 — CHANGES_REQUESTED** [review]

File: client/src/app/api/rest.ts (hunk 1 above)
Comment (gildub) [reply client/src/app/api/rest.ts:232]: @ibolton336, not really, working with `Ref` took me to make `id` not an option and this triggered a cascade of changes on the involved queries, which I aligned using axios as this is how we handle them.

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx (hunk 2 above)
Comment (gildub) [reply client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:122]: Same response

File: client/src/app/api/rest.ts (hunk 1 above)
Comment (ibolton336) [reply client/src/app/api/rest.ts:232]: Seems like some good changes here. Thanks.

File: client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx [hunk 3]
```diff
@@ -303,7 +303,7 @@ export const AnalysisWizard: React.FC<IAnalysisWizard> = ({
     { id, name },
     { prevId, prevName }
   ) => {
-    if (id && stepIdReached < id) setStepIdReached(id as number);
+    if (id && stepIdReached < (id as number)) setStepIdReached(id as number);
```
Comment (ibolton336) [inline client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx:306]: Why do we need this assertion?

Comment (gildub) [reply client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx:306]: Because Id is mandatory for that object anyway.

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx (hunk 2 above)
Comment (ibolton336) [reply client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:122]: Most of these changes are out of the scope of this PR. I know how it goes when you get rolling and want to make mass changes. But we should be looking to avoid this type of PR as a team. If we want to convert all of the rest.ts calls to remove the deprecated APIClient, then we should make a PR for that etc.

File: client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx (hunk 3 above)
Comment (gildub) [reply client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx:306]: More likely for the same reason we need at the end of that line too.

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx [hunk 4]
```diff
@@ -121,36 +170,22 @@ export const StakeholderGroupForm: React.FC<StakeholderGroupFormProps> = ({
         };
       });
 
-    const payload: StakeholderGroup = {
+    const payload: Omit<StakeholderGroup, "id"> = {
```
Comment (ibolton336) [inline client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:173]: Why is this necessary?

File: client/src/app/pages/controls/stakeholders/stakeholders.tsx [hunk 5]
```diff
@@ -61,8 +61,13 @@ export const Stakeholders: React.FC = () => {
...
 
-  const [isNewModalOpen, setIsNewModalOpen] = React.useState(false);
-  const [rowToUpdate, setRowToUpdate] = React.useState<Stakeholder>();
+  const [createUpdateModalState, setCreateUpdateModalState] = React.useState<
+    "create" | Stakeholder | null
+  >(null);
+  const isCreateUpdateModalOpen = createUpdateModalState !== null;
+  const stakeholderToUpdate =
+    createUpdateModalState !== "create" ? createUpdateModalState : null;
+
```
Comment (ibolton336) [inline client/src/app/pages/controls/stakeholders/stakeholders.tsx:70]: We should move the modal refactor work to a new PR and address it systematically if we want to do it.

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx (hunk 4 above)
Comment (gildub) [reply client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:173]: Because Id is always present but during creation.
This avoids to test for undefined Id when there is an ID.
I've updated description for more details

Comment (ibolton336) [reply client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:173]: Thats a good idea.

File: client/src/app/pages/controls/stakeholders/stakeholders.tsx (hunk 5 above)
Comment (gildub) [reply client/src/app/pages/controls/stakeholders/stakeholders.tsx:70]: The change with the Id, triggered a cascade of changes.
This is now aligned with our latest practice.
Sorry for the big PR :(

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx (hunk 2 above)
Comment (gildub) [reply client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:122]: I understand, and I agree with smaller PRs.
Again the cascade of change wasn't triggered to replace APIClient, but I made changes around then a aligned those.

File: client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx (hunk 4 above)
Comment (gildub) [reply client/src/app/pages/controls/stakeholder-groups/components/stakeholder-group-form.tsx:173]: Thanks.
I didn't mean to make so many changes but once I touched the ID it was a long journey to fix everything around it.
There are still some resources using optional ID but those weren't impacted so we can leave them as tech debt.

File: client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx (hunk 3 above)
Comment (ibolton336) [reply client/src/app/pages/applications/analysis-wizard/analysis-wizard.tsx:306]: Seems as though this is coming from patternfly allowing string or number for ids here.

File: client/src/app/pages/migration-waves/migration-waves.tsx [hunk 6]
```diff
@@ -244,8 +244,13 @@ export const MigrationWaves: React.FC = () => {
     return statuses as TicketStatus[];
   };
 
-  const aggregateTicketStatus = (val: TicketStatus) => {
+  const aggregateTicketStatus = (val: TicketStatus, startDate: string) => {
     const status = ticketStatusToAggreaate.get(val);
+    if (status === "Issues Created") {
+      const now = new Date();
```
Comment (ibolton336) [inline client/src/app/pages/migration-waves/migration-waves.tsx:250]: Lets try and use Dayjs here to be consistent with what was done in the wave create modal. Thx.

**Review by ibolton336 — APPROVED** [review]

File: client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx [hunk 7]
```diff
@@ -132,7 +169,7 @@ export const StakeholderForm: React.FC<StakeholderFormProps> = ({
         };
       });
 
-    const payload: Stakeholder = {
+    const payload: Omit<Stakeholder, "id"> = {
```
Comment (ibolton336) [inline client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx:172]: Seems like this was a good thought, but no reason we should have to remember to omit an ID every place we define an instance type for submitting payloads or setting the types in the rest.ts file. Seems like we could think of something better here.

**Review by ibolton336 — CHANGES_REQUESTED** [review]

File: client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx (hunk 7 above)
Comment (ibolton336) [reply client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx:172]: What is wrong with leaving the id optional and conditionally merging it to the payload object when we are updating?

Comment (gildub) [reply client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx:172]: There are no need to worry about Ids on any object with this approach.
The reason is because resources always have an ID but at creation time.
Instead until now we've been looking at it from the other end where Ids are optional because we don't know the value at creation time.  
The benefit mean certainty when handling such objects and not having to test if the object is not "undefined".

**Review by mturley — APPROVED** [review]
LGTM after discussion on Slack.

File: client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx (hunk 7 above)
Comment (ibolton336) [reply client/src/app/pages/controls/stakeholders/components/stakeholder-form.tsx]: Makes sense - thanks.

**Review by ibolton336 — APPROVED** [review]
